### PR TITLE
Improve swagger support for authenticators

### PR DESCRIPTION
**Pull request by airstandley**
BP-778
Issue #125

Makes it so Authenticator converters can define separate methods for getting security requirements and security definitions rather than making assumptions about the requirements based on the definition.

### CODE REVIEW

**Comment by Sytten** [comment]
I will do a full review of this (sick this week :( ), I think I might have a few suggestions for you since I built one from scratch.

File: flask_rebar/swagger_generation/authenticator_to_swagger.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    Abstract class for objects that convert Authenticator objects to
+    security JSONSchema.
+    """
+
+    AUTHENTICATOR_TYPE = None
+
+    def get_security_schemes(self, obj, context):
+        """
+        Get the security schemes for the provided Authenticator object
+        :param flask.authenticators.Authenticator obj:
```
Comment (Sytten) [inline flask_rebar/swagger_generation/authenticator_to_swagger.py]: Add a space to be consistent

File: flask_rebar/swagger_generation/authenticator_to_swagger.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        :param _Context context:
+        :rtype: dict
+        """
+        raise NotImplementedError()
+
+    def get_security_requirements(self, obj, context):
```
Comment (Sytten) [inline flask_rebar/swagger_generation/authenticator_to_swagger.py]: It took me a while to understand what you meant by requirements. I believe we should use `get_security_scopes` to stay consistent with the openapi documentation. Maybe also reference the part of their doc for that.

File: flask_rebar/swagger_generation/swagger_generator_v2.py [hunk 3]
```diff
@@ -116,20 +121,20 @@ def generate(
...
+        for authenticator in authenticators:
+            security_definitions.update(
+                self.authenticator_converter.get_security_schemes(authenticator)
+            )
+
+        default_security = []
+        for authenticator in registry.default_authenticators:
+            default_security.extend(
+                self.authenticator_converter.get_security_requirements(authenticator)
+            )
```
Comment (Sytten) [inline flask_rebar/swagger_generation/swagger_generator_v2.py]: So if I am not mistaken we only support the OR case:
```
    security:    # A OR B
      - A
      - B
```
Since OpenAPI supports more than that, I would like to make a case that we also support the combination of auth. Though this is out of scope for this PR.

File: flask_rebar/swagger_generation/authenticator_to_swagger.py (hunk 2 above)
Comment (airstandley) [reply flask_rebar/swagger_generation/authenticator_to_swagger.py]: Hmm, that's not good. 
I was trying to be consistent with the OpenApi documentation names, but the purpose of the methods needs to be as clear as possible.

Relevant parts I looked at:

> Definition/Scheme:
>https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#security-definitions-object
>https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#security-scheme-object
>https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#security-scheme-object

>Requirements:
>https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#security-requirement-object
>https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#security-requirement-object

@Sytten Do you think adding links  to reference those parts of the docs would be enough to remove confusion?

File: flask_rebar/swagger_generation/swagger_generator_v2.py (hunk 3 above)
Comment (airstandley) [reply flask_rebar/swagger_generation/swagger_generator_v2.py]: My thought was that since Authenticators are classes you can add as much complexity there as you want. (I don't think we ever want to limit the functionality of an authenticator to a validating a single security definition.)

Given that you had Authenticator A and Authenticator B, and for one endpoint you wanted to require A && B, you could already create an Authenticator C which combines A and B.

With this change you can now also define the correct swagger security objects for that endpoint, which is what was not possible before.

So while I agree that it might be nice to have better support for `A AND B`, this should mean that we at least can properly support it; even if it's a bit dirty.

Comment (airstandley) [reply flask_rebar/swagger_generation/swagger_generator_v2.py]: (But also I'm kicking myself for not considering that when working on multiple Authenticator support since it probably would not have been too much extra work to support.)

File: flask_rebar/swagger_generation/authenticator_to_swagger.py (hunk 2 above)
Comment (Sytten) [reply flask_rebar/swagger_generation/authenticator_to_swagger.py]: Yes, especially since we want users to extend those classes. I was using this doc: https://swagger.io/docs/specification/authentication/#scopes which doesn't mention the word security requirements anywhere :S

File: flask_rebar/swagger_generation/swagger_generator_v2.py (hunk 3 above)
Comment (Sytten) [reply flask_rebar/swagger_generation/swagger_generator_v2.py]: Agreed on that, maybe we should add documentation or examples on how to do that?

**Comment by airstandley** [comment]
Added better doc strings and recipe examples.

@Sytten would love to get your opinion.

**Comment by Sytten** [comment]
Sorry didnt see the message, looks very good to me!
